tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/NN.java

```diff
@@ -0,0 +1,379 @@
+   *     = (1 - z) * x + log(1 + exp(-x))
+   *     = x - x * z + log(1 + exp(-x))
+   * </pre>
+   *
+   * <p>For x < 0, to avoid overflow in exp(-x), we reformulate the above
+   *
+   * <pre>
+   *      x - x * z + log(1 + exp(-x))
+   *      = log(exp(x)) - x * z + log(1 + exp(-x))
+   *      = - x * z + log(1 + exp(x))
+   * </pre>
+   *
+   * <p>Hence, to ensure stability and avoid overflow, the implementation uses this equivalent
+   * formulation
+   *
+   * <pre>
+   *     max(x, 0) - x * z + log(1 + exp(-abs(x)))
+   * </pre>
+   *
+   * <p>`logits` and `labels` must have the same type and shape.
+   *
+   * @param scope The TensorFlow scope
+   * @param labels the labels
+   * @param logits the logits of type float32 or float64
+   * @param <T> the type of labels and logits
+   * @return the component-wise logistic losses.
+   */
+  @Endpoint(name = "sigmoidCrossEntropyWithLogits")
+  public static <T extends TNumber> Operand<T> sigmoidCrossEntropyWithLogits(
+      Scope scope, Operand<T> labels, Operand<T> logits) {
+    if (labels.asOutput().shape().numDimensions() != logits.asOutput().shape().numDimensions())
+      throw new IllegalArgumentException(
+          String.format(
+              "logits and labels must have the same shape (%s vs %s)",
+              labels.asOutput().shape().toString(), logits.asOutput().shape()));
+    Operand<T> zeros =
+        Cast.create(scope, ZerosLike.create(scope, logits), logits.asOutput().dataType());
+    Operand<TBool> cond = GreaterEqual.create(scope, logits, zeros);
+
+    Operand<T> relu_logits = Select.create(scope, cond, logits, zeros);
+    Operand<T> neg_abs_logits = Select.create(scope, cond, Neg.create(scope, logits), logits);
+    return Add.create(
+        scope,
+        Sub.create(scope, relu_logits, Mul.create(scope, logits, labels)),
+        Log1p.create(scope, Exp.create(scope, neg_abs_logits)));
+  }
+
+  /**
+   * Flattens logits' outer dimensions and keep its last dimension.
+   *
+   * @param scope the TensorFlow scope
+   * @param logits the logits
+   * @param <T> the type of logits
+   * @return the flattened logits
+   */
+  private static <T extends TNumber> Operand<T> flattenOuterDims(Scope scope, Operand<T> logits) {
+    Operand<TInt64> one = Constant.scalarOf(scope, 1L);
+
+    org.tensorflow.ndarray.Shape shape = logits.asOutput().shape();
+    int ndims = shape.numDimensions();
+    if (!shape.hasUnknownDimension()) {
+      long product = 1L;
+      boolean productValid = true;
+      for (int i = ndims - 2; i >= 0; i--) {
+        long d = shape.size(i);
+        if (d == org.tensorflow.ndarray.Shape.UNKNOWN_SIZE) {
+          productValid = false;
+          break;
+        }
+        product *= d;
+      }
+      if (productValid) {
+        org.tensorflow.ndarray.Shape outputShape = Shape.of(product, shape.size(ndims - 1));
+        return Reshape.create(scope, logits, Constant.vectorOf(scope, outputShape.asArray()));
```

Comment:
just `Constant.tensorOf(scope, shape.size(ndims - 1))` will do the job here

Reply:
I don't understand your comment.

This method, preserves the last dimension of shape, but does a product on all the dimensions but the last dimensions. The resulting dimension is a long[] of product, last size.
For example, shape [2,3,4] becomes [6,4].
This is a more concise way to do this though, getting rid of the intermediate shape.
```
return Reshape.create(scope, logits,
                        Constant.vectorOf(scope, new long[] { product, shape.size(-1)});
```

Reply:
I just changed it to use `Constant.arrayOf()`
```
return Reshape.create(
            scope, logits, Constant.arrayOf(scope, product, shape.size(-1)));
```

Reply:
Sorry, my comment was wrong as I've ignored the `product` value. What you did is right by using `Constant.arrayOf` is basically what I meant, i.e. there is no need to convert an array to a `Shape` if we are passing just back this array to `Constant`